### PR TITLE
Add surge and yaw PID controllers

### PR DESCRIPTION
**Pull request by whymarrh**
Refs #75 (which contains most of the discussion around the control system)
Refs #82 (which contains discussion about the controller implementation and the error functions)

This PR adds the surge and yaw controllers using the `PidController` introduced in #82.

### CODE REVIEW

File: projects/core/src/main/kotlin/core/Boat.kt [hunk 1]
```diff
@@ -22,17 +37,78 @@ class Boat(private val broadcast: Broadcast, private val propulsionSystem: Propu
...
+                    .scan(initialController, { cs, g -> cs.setGains(g.gains) })
+
+                val controlSystems = currentGps
+                    .withLatestFrom(controllers, { gps: GpsValue, controller: Controller -> Pair(controller, gps) })
+                    .map { (c, g) ->
+                        ControlSystem(c, { a: Position, b: Position ->
+                            val actualBearing = g.velocity.trueBearing.convert<Radian>(Radian)
+                            val desiredBearing = bearing(a, b).convert<Radian>(Radian)
+                            val error = (actualBearing - desiredBearing).value
+                            // Math.atan2(Math.sin(error), Math.cos(error))
```
Comment (whymarrh) [inline projects/core/src/main/kotlin/core/Boat.kt]: This (`Math.atan2`) should be used in place of the logic below

File: projects/core/src/main/kotlin/core/Boat.kt [hunk 2]
```diff
@@ -22,17 +37,78 @@ class Boat(private val broadcast: Broadcast, private val propulsionSystem: Propu
...
+            broadcast.valuesOfType<GpsFix>(),
+            broadcast.valuesOfType<GpsNavInfo>(),
+            { fix, nav -> Pair(fix, nav)}
+        )
+            .map { (fix, nav) -> GpsValue.from(nav, fix) }
+            .filter { it != null }
+            .map { it!! }
+
+        val config = broadcast.valuesOfType<BoatConfig>()
+            .startWith(BoatConfig(SurgeGains(0.0, 0.0, 0.0), 1.0, YawGains(-0.5, 0.0, 0.0), 1.0))
```
Comment (whymarrh) [inline projects/core/src/main/kotlin/core/Boat.kt]: These gains should default to 1s

File: projects/core/src/main/kotlin/core/Boat.kt [hunk 3]
```diff
@@ -22,17 +37,78 @@ class Boat(private val broadcast: Broadcast, private val propulsionSystem: Propu
...
             { fix, nav -> Pair(fix, nav)}
         )
 
+        val currentGps = Observable.combineLatest(
+            broadcast.valuesOfType<GpsFix>(),
+            broadcast.valuesOfType<GpsNavInfo>(),
+            { fix, nav -> Pair(fix, nav)}
+        )
+            .map { (fix, nav) -> GpsValue.from(nav, fix) }
+            .filter { it != null }
```
Comment (whymarrh) [inline projects/core/src/main/kotlin/core/Boat.kt]: This should be replaced with `mapNotNull`

File: projects/core/src/main/kotlin/core/Boat.kt [hunk 4]
```diff
@@ -1,10 +1,25 @@
...
+import core.values.ControlMode
 import core.values.GpsValue
 import core.values.Motion
+import core.values.Position
+import core.values.SurgeGains
+import core.values.Waypoint
+import core.values.YawGains
 import gps.GpsFix
 import gps.GpsNavInfo
+import log.Log
```
Comment (whymarrh) [inline projects/core/src/main/kotlin/core/Boat.kt]: Nit: this is unused

File: projects/core/src/main/kotlin/core/Boat.kt [hunk 5]
```diff
@@ -22,17 +37,78 @@ class Boat(private val broadcast: Broadcast, private val propulsionSystem: Propu
...
+                    .onBackpressureDrop()
+                    .compose(ControlSystemTransformer(waypoint.position, controlSystems, clock))
+            })
+        val speeds = broadcast.valuesOfType<ControlMode>()
+            .startWith(ControlMode.MANUAL)
+            .switchMap({
+                when (it) {
+                    ControlMode.MANUAL   -> broadcast.valuesOfType<Motion>()
+                    ControlMode.WAYPOINT -> Observable.combineLatest(
+                        surge.map({ -0.75 })
```
Comment (whymarrh) [inline projects/core/src/main/kotlin/core/Boat.kt]: This should be removed as it's incorrect in the general case

File: projects/core/src/main/kotlin/core/Boat.kt [hunk 6]
```diff
@@ -22,17 +37,78 @@ class Boat(private val broadcast: Broadcast, private val propulsionSystem: Propu
...
         val positions = Observable.combineLatest(
             broadcast.valuesOfType<GpsFix>(),
             broadcast.valuesOfType<GpsNavInfo>(),
             { fix, nav -> Pair(fix, nav)}
         )
 
+        val currentGps = Observable.combineLatest(
+            broadcast.valuesOfType<GpsFix>(),
+            broadcast.valuesOfType<GpsNavInfo>(),
+            { fix, nav -> Pair(fix, nav)}
```
Comment (whymarrh) [inline projects/core/src/main/kotlin/core/Boat.kt]: This combiner function could be replaced with `GpsValue.from(nav, fix)` from below

File: projects/core/src/main/kotlin/core/Boat.kt [hunk 7]
```diff
@@ -22,17 +37,78 @@ class Boat(private val broadcast: Broadcast, private val propulsionSystem: Propu
     private val killSwitch: Subject<Void, Void> = PublishSubject.create()
 
     fun start(io: Scheduler, clock: Scheduler) {
-        val speeds = broadcast.valuesOfType<Motion>()
-            .startWith(Motion(0.0, 0.0))
-            .map { speed(it) }
-            .onBackpressureLatest()
```
Comment (whymarrh) [inline projects/core/src/main/kotlin/core/Boat.kt]: This backpressure operator is missing from the new switch map

**Comment by whymarrh** [comment]
@arandell93 can I get a 2nd pair of eyes on the error functions:

Surge (distance):

```kotlin
fun (a: Position, b: Position): Double {
    return distance(a, b).value
}
```

Yaw (bearing):

```kotlin
fun (a: Position, b: Position): Double {
    val actualBearing = g.velocity.trueBearing.convert<Radian>(Radian)
    val desiredBearing = bearing(a, b).convert<Radian>(Radian)
    val error = (actualBearing - desiredBearing).value
    return Math.atan2(Math.sin(error), Math.cos(error))
}
```

We've already discussed this in #82 and #75 but I want to double-check these. I know we'll revisit these when we get the new GPS on the go and start tuning the controller.

**Comment by arandell93** [comment]
Those look right to me!

**Pull request merged**